onnxruntime/core/xnnpack/optimizer/xnnpack_transformer.cc

```diff
@@ -0,0 +1,182 @@
+    Node* node_p = main_graph.GetNode(ni);
+    if (node_p == nullptr)
+      continue;
+    Node& nodeRef = *node_p;
+    ProtoHelperNodeContext nc(nodeRef);
+    int64_t group = 1;
+    OpNodeProtoHelper info(&nc);
+    auto X_input = info.GetInputType(0);
+    auto weight_input = info.GetInputType(1);
+    TensorShape weight_shape = utils::GetTensorShapeFromTensorShapeProto(weight_input->tensor_type().shape());
+    TensorShape X_shape = utils::GetTensorShapeFromTensorShapeProto(X_input->tensor_type().shape());
+    if (weight_shape.NumDimensions() != 4) continue;
+    if (group != 1 && group != X_shape[3]) continue;
+    for (size_t i = 0; i != weight_shape.NumDimensions(); ++i) {
+      if (weight_shape[i] <= 0) continue;
+    }
+    modified = true;
+    // const_cast
+    Node* conv_node = main_graph.GetNode(nodeRef.Index());
+    NodeArg* bias_node_arg = nullptr;
+    const bool has_bias = conv_node->InputDefs().size() >= 3;
+    if (!has_bias) {
+      ::ONNX_NAMESPACE::TensorProto bias_tensor;
+      int64_t bias_size = weight_shape[0];
+      std::vector<float> bias_data(bias_size, 0.0f);
+      bias_tensor.mutable_float_data()->Add(bias_data.begin(), bias_data.end());
+      bias_tensor.mutable_dims()->Add(bias_size);
+      std::string bias_tensor_name = main_graph.GenerateNodeArgName(conv_node->Name() + "_bias");
+      bias_tensor.set_name(bias_tensor_name);
+      bias_tensor.set_data_type(ONNX_NAMESPACE::TensorProto_DataType_FLOAT);
+      bias_node_arg = &graph_utils::AddInitializer(main_graph, bias_tensor);
+    }
+    std::string auto_pad_str;
+    ORT_RETURN_IF_ERROR(info.GetAttr<std::string>("auto_pad", &auto_pad_str));
+    ORT_RETURN_IF_ERROR(info.GetAttr<int64_t>("group", &group));
+    // group == 1 || group  == input / output channel count
+    // For now we assume input channel count isn't 1, so that group count != input/output channel count
+    bool is_depthwise = X_shape[3] != 1 && group == X_shape[3];
+
+    InOutDefSlot src_slot{ArgType::kInput, 1};
+    // Append a single slot to dest. As the dest is empty, it will be the first one.
+    ValueMoveInfo value_move_info(src_slot, ArgType::kInput, false, false);
+
+    if (conv_node == nullptr) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Couldn't find the node with index=", nodeRef.Index());
+    }
+    if (conv_node->InputDefs().size() < 2) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Expect at least 2 inputs, got ", conv_node->InputDefs().size());
+    }
+
+    // bool is_pointwise = weight_shape[2] == 1 && weight_shape[3] == 1;
+    // if (!is_pointwise) {
+    std::vector<int64_t> input_perm = is_depthwise ? std::vector<int64_t>{1, 2, 3, 0} : std::vector<int64_t>{0, 2, 3, 1};
+    std::string output_name = main_graph.GenerateNodeArgName("trans");
+    NodeArg& transpose_output = main_graph.GetOrCreateNodeArg(output_name, nullptr);
+
+    Node& dest_node = main_graph.AddNode("", "Transpose", "", {}, {&transpose_output}, nullptr, kOnnxDomain);
```

Comment:
which not takes its output as input?

Reply:
Two lines below, MoveInputOutput, creates the input for this node.